keeper/component/dispatcher/query_server.go

```diff
@@ -0,0 +1,192 @@
+	return queryServer{Dispatcher: d}
+}
+
+func (q queryServer) DispatchedCounts(
+	ctx context.Context,
+	req *dispatchertypes.QueryDispatchedCountsRequest,
+) (*dispatchertypes.QueryDispatchedCountsResponse, error) {
+	if req == nil {
+		return nil, sdkerrors.ErrInvalidRequest
+	}
+
+	sourceID, err := core.NewCrossChainID(req.SourceProtocolId, req.SourceCounterpartyId)
+	if err != nil {
+		return nil, errorsmod.Wrapf(err, "error creating source cross-chain ID")
+	}
+
+	destID, err := core.NewCrossChainID(req.DestinationProtocolId, req.DestinationCounterpartyId)
+	if err != nil {
+		return nil, errorsmod.Wrapf(err, "error creating destination cross-chain ID")
+	}
+
+	if !q.HasDispatchedCounts(ctx, &sourceID, &destID) {
+		return nil, fmt.Errorf(
+			"dispatched counts do not exist for source ID %s and destination ID %s",
+			sourceID.String(),
+			destID.String(),
+		)
+	}
+
+	counts := q.GetDispatchedCounts(ctx, &sourceID, &destID)
+
+	return &dispatchertypes.QueryDispatchedCountsResponse{
+		Counts: []*dispatchertypes.DispatchCountEntry{counts},
+	}, nil
+}
+
+func (q queryServer) DispatchedCountsByDestinationProtocolID(
+	ctx context.Context,
+	req *dispatchertypes.QueryDispatchedCountsByProtocolIDRequest,
+) (*dispatchertypes.QueryDispatchedCountsResponse, error) {
+	if req == nil {
+		return nil, sdkerrors.ErrInvalidRequest
+	}
+
+	if err := req.ProtocolId.Validate(); err != nil {
+		return nil, errorsmod.Wrapf(err, "invalid protocol ID")
+	}
+
+	counts := q.GetDispatchedCountsByDestinationProtocolID(ctx, req.ProtocolId)
+
+	return &dispatchertypes.QueryDispatchedCountsResponse{
+		Counts: counts,
+	}, nil
+}
+
+func (q queryServer) DispatchedCountsBySourceProtocolID(
+	ctx context.Context,
+	req *dispatchertypes.QueryDispatchedCountsByProtocolIDRequest,
+) (*dispatchertypes.QueryDispatchedCountsResponse, error) {
+	if req == nil {
+		return nil, sdkerrors.ErrInvalidRequest
+	}
+
+	if err := req.ProtocolId.Validate(); err != nil {
+		return nil, errorsmod.Wrapf(err, "invalid protocol ID")
+	}
+
+	counts := q.GetDispatchedCountsBySourceProtocolID(ctx, req.ProtocolId)
+
+	return &dispatchertypes.QueryDispatchedCountsResponse{
+		Counts: counts,
+	}, nil
+}
+
+func (q queryServer) DispatchedAmounts(
+	ctx context.Context,
+	req *dispatchertypes.QueryDispatchedAmountsRequest,
+) (*dispatchertypes.QueryDispatchedAmountsResponse, error) {
+	if req == nil {
+		return nil, sdkerrors.ErrInvalidRequest
+	}
+
+	sourceID, err := core.NewCrossChainID(req.SourceProtocolId, req.SourceCounterpartyId)
+	if err != nil {
+		return nil, errorsmod.Wrapf(err, "error creating source cross-chain ID")
+	}
+
+	destID, err := core.NewCrossChainID(req.DestinationProtocolId, req.DestinationCounterpartyId)
+	if err != nil {
+		return nil, errorsmod.Wrapf(err, "error creating destination cross-chain ID")
+	}
+
+	if req.Denom == "" {
+		return nil, errorsmod.Wrapf(
+			core.ErrEmptyString,
+			"error querying an empty string token denom",
+		)
+	}
+
```

Comment:
nit but I think this could already be handled before building the cross-chain IDs so we don't waste operations if there is no denom passed